### ISSUE TITLE
Gradient checkpointing

### ISSUE THREAD

**Issue by f4at** [tensorflow/addons#2644]
**Describe the feature and the current behavior/state.**
By checkpointing nodes in the computation graph defined by your model, and recomputing the parts of the graph in-between those nodes during backpropagation, it is possible to calculate this gradient at reduced memory cost, without sacrificing a lot of performance. It's a lot better than using gradient accumulation.

**Relevant information**
- Are you willing to contribute it (yes/no): no
- Are you willing to maintain it going forward? (yes/no): no
- Is there a relevant academic paper? (if so, where): https://arxiv.org/abs/1604.06174v2
- Does the relavent academic paper exceed 50 citations? (yes/no): no
- Is there already an implementation in another framework? (if so, where):  Official implementation in tf1: https://github.com/cybertronai/gradient-checkpointing and a basic implementation in tf2 eager mode; https://github.com/davisyoshida/tf2-gradient-checkpointing 
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)**
I'm not sure.

**Who will benefit with this feature?**
People who want to train big models (e.g: GPT-2, etc..) but don't have enough memory and don't want to sacrifice performance either.

**Any other info.**
The official implementation works only for tf1 and the basic implementation works in tf2 eager mode only which is slow and it is very basic and is not polished so it doesn't give the maximal performance.
